### PR TITLE
Added static files serving.

### PR DESCRIPTION
**Pull request by dkorolev**

### CODE REVIEW

**Comment by dkorolev** [comment]
Addresses issue https://github.com/KnowSheet/Bricks/issues/71

**Comment by dkorolev** [comment]
@sompylasar @deathbaba : PTAL

File: net/api/test.cc [hunk 1]
```diff
@@ -494,3 +495,32 @@ TEST(HTTPAPI, UserAgent) {
...
+  const std::string dir = FLAGS_net_api_test_tmpdir + "/static";
+  FileSystem::MkDir(dir, FileSystem::MkDirParameters::Silent);
+  FileSystem::WriteStringToFile("<h1>HTML</h1>", FileSystem::JoinPath(dir, "file.html").c_str());
+  FileSystem::WriteStringToFile("This is text.", FileSystem::JoinPath(dir, "file.txt").c_str());
+  FileSystem::WriteStringToFile("And this: PNG", FileSystem::JoinPath(dir, "file.png").c_str());
+  FileSystem::MkDir(FileSystem::JoinPath(dir, "subdir_to_ignore"), FileSystem::MkDirParameters::Silent);
+  HTTP(FLAGS_net_api_test_port).ServeStaticFilesFrom(dir);
+  EXPECT_EQ("<h1>HTML</h1>", HTTP(POST(Printf("localhost:%d/file.html", FLAGS_net_api_test_port))).body);
+  EXPECT_EQ("This is text.", HTTP(POST(Printf("localhost:%d/file.txt", FLAGS_net_api_test_port))).body);
+  EXPECT_EQ("And this: PNG", HTTP(POST(Printf("localhost:%d/file.png", FLAGS_net_api_test_port))).body);
```
Comment (sompylasar) [inline net/api/test.cc]: No test for "subdir_to_ignore" not given out to a request.

File: net/api/test.cc [hunk 2]
```diff
@@ -494,3 +495,32 @@ TEST(HTTPAPI, UserAgent) {
...
+  HTTP(FLAGS_net_api_test_port).ResetAllHandlers();
+  FileSystem::MkDir(FLAGS_net_api_test_tmpdir, FileSystem::MkDirParameters::Silent);
+  const std::string dir = FLAGS_net_api_test_tmpdir + "/static";
+  FileSystem::MkDir(dir, FileSystem::MkDirParameters::Silent);
+  FileSystem::WriteStringToFile("<h1>HTML</h1>", FileSystem::JoinPath(dir, "file.html").c_str());
+  FileSystem::WriteStringToFile("This is text.", FileSystem::JoinPath(dir, "file.txt").c_str());
+  FileSystem::WriteStringToFile("And this: PNG", FileSystem::JoinPath(dir, "file.png").c_str());
+  FileSystem::MkDir(FileSystem::JoinPath(dir, "subdir_to_ignore"), FileSystem::MkDirParameters::Silent);
+  HTTP(FLAGS_net_api_test_port).ServeStaticFilesFrom(dir);
+  EXPECT_EQ("<h1>HTML</h1>", HTTP(POST(Printf("localhost:%d/file.html", FLAGS_net_api_test_port))).body);
```
Comment (sompylasar) [inline net/api/test.cc]: These must be only `GET` requests.

File: net/api/impl/posix_server.h [hunk 3]
```diff
@@ -190,6 +192,30 @@ class HTTPServerPOSIX final {
     handlers_.erase(path);
   }
 
+  struct StaticFileServer {
+    std::string body;
+    std::string content_type;
+    explicit StaticFileServer(const std::string& body, const std::string& content_type)
+        : body(body), content_type(content_type) {}
+    void operator()(Request r) { r.connection.SendHTTPResponse(body, HTTPResponseCode.OK, content_type); }
```
Comment (sompylasar) [inline net/api/impl/posix_server.h]: Only `GET` requests must be served, other should response with `Method Not Allowed`

**Comment by sompylasar** [comment]
@dkorolev PTAL, added some comments.

**Comment by deathbaba** [comment]
Needs rebasing

**Comment by dkorolev** [comment]
Why rebasing and not merging?

On Monday, February 23, 2015, Alex Zolotarev <notifications@github.com>
wrote:

> Needs rebasing
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/KnowSheet/Bricks/pull/86#issuecomment-75629221>.
>


-- 
Thanks,
Dima

http://dimakorolev.com/ | http://www.linkedin.com/in/dimakorolev |
dmitry.korolev@gmail.com | +1 (312) 593 2783

File: net/api/impl/posix_server.h (hunk 3 above)
Comment (deathbaba) [inline net/api/impl/posix_server.h]: Why GET only? POST files also can serve static content if necessary.

File: net/api/impl/posix_server.h [hunk 4]
```diff
@@ -190,6 +192,30 @@ class HTTPServerPOSIX final {
...
+  };
+
+  void ServeStaticFilesFrom(const std::string& dir, const std::string& route_prefix = "/") {
+    // TODO(dkorolev): Add a scoped version of registerers.
+    FileSystem::ScanDir(dir, [this, &dir, &route_prefix](const std::string& file) {
```
Comment (deathbaba) [inline net/api/impl/posix_server.h]: Scandir is not recursive, right?

File: net/api/impl/posix_server.h (hunk 3 above)
Comment (dkorolev) [inline net/api/impl/posix_server.h]: I'll go with John's suggestion of GET only. It sounds natural to me that
POST-ing into static routes should result in 403 Method Not Allowed.

On Monday, February 23, 2015, Alex Zolotarev <notifications@github.com>
wrote:

> In net/api/impl/posix_server.h
> <https://github.com/KnowSheet/Bricks/pull/86#discussion_r25201074>:
>
> > @@ -190,6 +192,30 @@ class HTTPServerPOSIX final {
> >      handlers_.erase(path);
> >    }
> >
> > +  struct StaticFileServer {
> > +    std::string body;
> > +    std::string content_type;
> > +    explicit StaticFileServer(const std::string& body, const std::string& content_type)
> > +        : body(body), content_type(content_type) {}
> > +    void operator()(Request r) { r.connection.SendHTTPResponse(body, HTTPResponseCode.OK, content_type); }
>
> Why GET only? POST files also can serve static content if necessary.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/KnowSheet/Bricks/pull/86/files#r25201074>.
>


-- 
Thanks,
Dima

http://dimakorolev.com/ | http://www.linkedin.com/in/dimakorolev |
dmitry.korolev@gmail.com | +1 (312) 593 2783

**Comment by deathbaba** [comment]
Merge button is disabled for me.

**Comment by dkorolev** [comment]
@deathbaba @sompylasar : Thanks for the review. PTAL.

File: net/api/impl/posix_server.h (hunk 4 above)
Comment (dkorolev) [inline net/api/impl/posix_server.h]: Yep.

**Pull request merged**